drivers: pwm: add pwm driver support in NPCX7 series

MulinChao opened the pull request:
In npcx7 series, there're 8 Pulse Width Modulator (PWM) modules and each one support generating a single 16-bit PWM output. A 16-bit clock prescaler (PRSCn) and a 16-bit counter (CTRn) determine the cycle time, the minimal possible pulse width, and the duty-cycle steps.

Beside introducing pwm driver for Nuvoton NPCX series, this PR also includes:

1. Add PWM device tree declarations.
2. Zephyr PWM api implementation.
3. Add aliases in npcx7m6fb_evb board device tree file for supporting
   samples/basic/blinky_pwm application and pwm test suites

MulinChao committed:
drivers: pwm: add pwm driver support in NPCX7 series

In npcx7 series, there're 8 Pulse Width Modulator (PWM) modules and each
one support generating a single 16-bit PWM output. A 16-bit clock
prescaler (PRSCn) and a 16-bit counter (CTRn) determine the cycle time,
the minimal possible pulse width, and the duty-cycle steps.

Beside introducing pwm driver for Nuvoton NPCX series, this CL also
includes:

1. Add PWM device tree declarations.
2. Zephyr PWM api implementation.
3. Add aliases in npcx7m6fb_evb board device tree file for supporting
   samples/basic/blinky_pwm application and pwm test suites

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

sjg20 commented on drivers/pwm/pwm_npcx.c:
```diff
@@ -0,0 +1,213 @@
+	 * Then prescaler = ceil (period_cycles / pwm_max_period_cycles)
+	 */
+	prescaler = ceiling_fraction(period_cycles, NPCX_PWM_MAX_PERIOD_CYCLES);
+
```
drop blank line

MulinChao replied:
Done

sjg20 commented on drivers/pwm/pwm_npcx.c:
```diff
@@ -0,0 +1,213 @@
+
+	/* Turn on device clock first and get source clock freq of it. */
+	if (clock_control_on(clk_dev,
+		(clock_control_subsys_t *) &config->clk_cfg) != 0) {
```
drop space before &

MulinChao replied:
Done.

sjg20 commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -21,6 +21,24 @@
+ * registers. For example, a word register access might break into two byte
+ * register accesses by adding 'packed' attribute.
+ */
+#define NPCX_REG_WORD_ACCESS_CHECK(reg, val) { \
```
Please add a comment for each macro so we know what the args are supposed to me and what it does

MulinChao replied:
Done.

sjg20 commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -308,4 +326,31 @@ struct gpio_reg {
+ */
+struct pwm_reg {
+	/* 0x000: Clock Prescaler */
+	volatile uint16_t PRSC;
```
My opinion of the upper-case register names has not improved

MulinChao replied:
Here is our discussion for this in the other PR. https://github.com/zephyrproject-rtos/zephyr/pull/27578#discussion_r474971026. Currently, it doesn't violate the code style of Zephyr. I still strongly suggest keeping them the same as our datasheet possibly. The other important reason is I want to keep the same style as our HAL did. Or it makes me  hard to maintain updates between Zephry and NPCX HAL drivers

jettr replied:
I think we should keep is consistent with other Nuvoton code for now. If we want to change it, we should change it all of the register definitions for Nuvoton.

sjg20 commented on soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series:
```diff
@@ -35,6 +35,12 @@ config GPIO_NPCX
 	help
 	  Enable support for NPCX GPIO driver.
 
+config PWM_NPCX
```
How come this option is defined in two places?

MulinChao replied:
The first definition in Kconfig.npcx is introducing this variable. And the second definition makes this variable turn on as default if chip series is npcx 7. You can refer the similar declarations in https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/pwm/Kconfig.xec#L6 and https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/microchip_mec/mec1501/Kconfig.defconfig.mec1501hsz#L46.

sjg20 commented on soc/arm/nuvoton_npcx/npcx7/soc.c:
```diff
@@ -46,3 +46,9 @@ NPCX_REG_OFFSET_CHECK(uart_reg, UFRCTL, 0x026);
+NPCX_REG_SIZE_CHECK(pwm_reg, 0x00E);
+NPCX_REG_OFFSET_CHECK(pwm_reg, PWMCTL, 0x004);
+NPCX_REG_OFFSET_CHECK(pwm_reg, DCR, 0x006);
+NPCX_REG_OFFSET_CHECK(pwm_reg, PWMCTLEX, 0x00C);
```
lower-case hex

MulinChao replied:
Here is our discussion for this in the other PR. https://github.com/zephyrproject-rtos/zephyr/pull/27578#discussion_r474969730. The reason that I use 12-bits offset is that all npcx module register sizes won't exceed 4KB although we reserved 8K for them. If you think it's a crucial code style must follow, I will add one more commit in this PR to fix all of them in soc.c.

jettr commented on drivers/pwm/pwm_npcx.c:
```diff
@@ -0,0 +1,213 @@
+/* PWM local functions */
+static void pwm_npcx_configure(const struct device *dev, int clk_bus)
+{
+	struct pwm_reg *inst = HAL_INSTANCE(dev);
```
nit: here and all functions we can add another const modifier

```suggestion
	struct pwm_reg *const inst = HAL_INSTANCE(dev);
```

MulinChao replied:
Done.

jettr replied:
I was thinking that you would also update the other variables pointers as well to have the `const` modifier, like `config`, `data`, and `clk_dev` in other functions.

If you have to make other changes in this file, you can make these updates. If there are no other required file updates, then you can leave it the way it is.

MulinChao replied:
Actually, I think it's a good discipline for this driver and following npcx drivers.  And I will update it to previous drivers in a new commit for consistency.

pabigot replied:
This is a coding style decision that we really ought to get consensus on before making wholesale changes, because if we're going to do it we should do it consistently, and following a rule.  Historically only a few drivers qualified the pointers that they copied out of immutable structures to local (const) variables.  Should every variable definition that the developer/maintainer believes will not be changed be marked const?

Making local variables `const` does clarify intent, but is there a clear risk to standardizing on unqualified local "variables"?

It's not a particularly big deal (and I do this frequently myself in C++), but we use Coccinelle to automate code transformations, and when those transformations involve adding or removing `const` qualifiers the presence of extra `const` qualifiers makes it difficult to get the transformation rules correct.

I've also seen this in API function declarations to indicate the parameter variable will not be changed, which I do object to.

jettr replied:
These are new files we are adding, so I think it is okay to suggest and make the changes here. I think within a file, we should definitely be self consistent; that is why I recommend the other variables in this file to also add the `const` modifier.

I like the idea of including this in Zephyrs style (or proposing it at least). I do wonder if there is a coverity option we can turn on somewhere.

I think use of `const` should be used when it is possible as it can prevent accidental coding mistakes as well as self-document the variables purpose better.

jettr commented on drivers/pwm/pwm_npcx.c:
```diff
@@ -0,0 +1,213 @@
+	struct pwm_reg *inst = HAL_INSTANCE(dev);
+
+	/*
+	 * NPCX PWM modulee mix byte and word registers together. Make sure
```
```suggestion
	 * NPCX PWM module mixes byte and word registers together. Make sure
```

MulinChao replied:
Done.

jettr commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -21,6 +21,24 @@
+ * register accesses by adding 'packed' attribute.
+ */
+#define NPCX_REG_WORD_ACCESS_CHECK(reg, val) { \
+		uint16_t dummy = reg; \
```
nit: let's try not to use word dummy as it can related to intelligence. 
```suggestion
		uint16_t placeholder = reg; \
```

MulinChao replied:
Done.

jettr reviewed (changes requested):
Looks good, just a few minor comments

MulinChao commented on drivers/pwm/pwm_npcx.c:
```diff
@@ -0,0 +1,213 @@
+	struct pwm_reg *inst = HAL_INSTANCE(dev);
+
+	/* Disable PWM for module configuration first */
+	inst->PWMCTL &= ~(BIT(NPCX_PWMCTL_PWR));
```
Done.

MulinChao commented on drivers/pwm/pwm_npcx.c:
```diff
@@ -0,0 +1,213 @@
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const uint8_t   alts_size;
```
Done.

MulinChao commented on drivers/pwm/pwm_npcx.c:
```diff
@@ -0,0 +1,213 @@
+	if (clk_bus == NPCX_CLOCK_BUS_LFCLK)
+		inst->PWMCTL |= BIT(NPCX_PWMCTL_CKSEL);
+	else
+		inst->PWMCTL &= ~(BIT(NPCX_PWMCTL_CKSEL));
```
Done.

MulinChao commented on drivers/pwm/Kconfig.npcx:
```diff
@@ -0,0 +1,12 @@
+	bool "Nuvoton NPCX embedd controller (EC) PWM driver"
+	depends on SOC_FAMILY_NPCX
+	help
+	  This option enables the PWM driver for NPCX family of
```
It has been mentioned in the commit message. I also checked help messagesi n the other Kconfig file. Most of them are brief and don't mention how many channels it provides.  Is it a common case to add detailed help message here?

MulinChao committed:
drivers: npcx: Add const modifier for hal instances and so on.

Add const modifier for hal instances, clock devices pointer, and module
base address in npcx drivers to prevent driver functions change them
unexpectedly.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

MulinChao commented:
Hi All:
This PR is pending for a while. Do you have any comment for it? Thanks.

pabigot commented:
AFAICT it's ready to go in, but will have to wait for the 2.5.0 merge window to open after the release.

MulinChao commented:
> AFAICT it's ready to go in, but will have to wait for the 2.5.0 merge window to open after the release.

Got it. Thanks for your reply.

carlescufi commented:
@MulinChao can you please rebase it?

MulinChao commented:
> @MulinChao can you please rebase it?

@carlescufi Sorry for the late reply. It's done. Thank you.

MaureenHelm commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Nuvoton Technology Corporation.
+yright (c) 2020 Nuvoton Technology Corporation.
```
Mistake?

MulinChao replied:
Oops. Thanks for reminding.

MulinChao replied:
Done.
